File: lib/extended/cells.js
'use strict';

const { isInterior, cannotSleep, getCellName, isDead } = require('../util');
const ignore = require('../../data/ignore').extended;

const IGNORE_INHABITANTS = new Set(ignore.inhabitants.map(s => s.toLowerCase()));
const IGNORE_PATH_GRIDS = new Set(ignore.pathgrids.map(s => s.toLowerCase()));

const INHABITANTS = new Set();
const PATH_GRIDS = new Set();
const CELLS = [];
const MIN_INHABITANTS = 3;

module.exports = {
	onRecord(record, last, recordId) {
		if(record.type === 'PathGrid') {
			if(record.cell) {
				PATH_GRIDS.add(record.cell.toLowerCase());
			}
		} else if(record.type === 'Cell') {
			if(last && isInterior(record) && record.references.length > 1) {
				CELLS.push(record);
				if(!cannotSleep(record) && !IGNORE_INHABITANTS.has(recordId)) {
					const count = record.references.filter(ref => INHABITANTS.has(ref.id.toLowerCase())).length;
					if(count < MIN_INHABITANTS) {
						console.log(record.type, getCellName(record), 'contains', count, 'NPCs or creatures');
					}
				}
			}
		} else if(record.type === 'LevelledCreature') {
			INHABITANTS.add(recordId);
		} else if(['Creature', 'Npc'].includes(record.type) && !isDead(record)) {
			INHABITANTS.add(recordId);
		}
	},
	onEnd() {
		CELLS.forEach(cell => {
			const id = cell.id.toLowerCase();
			if(!PATH_GRIDS.has(id) && !IGNORE_PATH_GRIDS.has(id)) {
				console.log(cell.type, getCellName(cell), 'is missing a path grid');
			}
		});
	}
};

File: src/extended/cells.rs
use std::collections::{HashMap, HashSet};

use clap::ArgMatches;
use tes3::esp::{EditorId, TES3Object};

use crate::{
    context::Context,
    util::{cannot_sleep, Actor},
};

use super::ExtendedHandler;

pub struct CellValidator {
    inhabitants: HashSet<String>,
    pathgrids: HashSet<String>,
    cells: Vec<(String, String)>,
    min_inhabitants: usize,
    regions: HashMap<(i32, i32), String>,
    changed: HashSet<(i32, i32)>,
}

fn is_region_change(x: i32, y: i32, changed: &HashSet<(i32, i32)>) -> bool {
    let mut surrounding = 0;
    for dx in -1..=1 {
        for dy in -1..=1 {
            if dx == 0 && dy == 0 {
                continue;
            }
            if changed.contains(&(x + dx, y + dy)) {
                surrounding += 1;
                if surrounding > 1 || changed.contains(&(x + dx * 2, y + dy * 2)) {
                    return false;
                }
            }
        }
    }
    true
}

impl ExtendedHandler for CellValidator {
    fn on_record(&mut self, _: &Context, record: &TES3Object, _: &str, last: bool) {
        match record {
            TES3Object::PathGrid(pathgrid) => {
                if !pathgrid.cell.is_empty() {
                    self.pathgrids.insert(pathgrid.cell.to_ascii_lowercase());
                }
            }
            TES3Object::Cell(cell) => {
                if last
                    && cell.is_interior()
                    && cell.references.len() > 1
                    && !cell.name.starts_with("T_")
                {
                    self.cells
                        .push((cell.name.to_ascii_lowercase(), cell.editor_id().into()));
                    if !cannot_sleep(cell) {
                        let count = cell
                            .references
                            .iter()
                            .filter(|(_, r)| self.inhabitants.contains(&r.id.to_ascii_lowercase()))
                            .count();
                        if count < self.min_inhabitants {
                            println!(
                                "Cell {} contains {} NPCs or creatures",
                                cell.editor_id(),
                                count
                            );
                        }
                    }
                }
                if cell.is_exterior() {
                    if last {
                        let current = self.regions.get(&cell.data.grid);
                        if let Some(region) = current {
                            if !cell
                                .region
                                .as_ref()
                                .is_some_and(|r| r.eq_ignore_ascii_case(region))
                            {
                                self.changed.insert(cell.data.grid);
                            } else {
                                return;
                            }
                        } else {
                            return;
                        }
                    }
                    if let Some(region) = &cell.region {
                        self.regions.insert(cell.data.grid, region.clone());
                    } else {
                        self.regions.remove(&cell.data.grid);
                    }
                }
            }
            TES3Object::LeveledCreature(_) => {
                self.inhabitants
                    .insert(record.editor_id_ascii_lowercase().into_owned());
            }
            TES3Object::Creature(c) => {
                if !c.is_dead() {
                    self.inhabitants
                        .insert(record.editor_id_ascii_lowercase().into_owned());
                }
            }
            TES3Object::Npc(n) => {
                if !n.is_dead() {
                    self.inhabitants
                        .insert(record.editor_id_ascii_lowercase().into_owned());
                }
            }
            _ => {}
        }
    }

    fn on_end(&mut self) {
        for (id, name) in &self.cells {
            if !self.pathgrids.contains(id) {
                println!("Cell {} is missing a path grid", name);
            }
        }
        for grid in &self.changed {
            if is_region_change(grid.0, grid.1, &self.changed) {
                let region = self.regions.get(grid).map_or("None", String::as_ref);
                println!(
                    "Cell ({}, {}) had its region changed to {}",
                    grid.0, grid.1, region
                );
            }
        }
    }
}

impl CellValidator {
    pub fn new(args: &ArgMatches) -> Self {
        let min_inhabitants = *args.get_one::<usize>("mininhabitants").unwrap();
        Self {
            inhabitants: HashSet::new(),
            pathgrids: HashSet::new(),
            cells: Vec::new(),
            min_inhabitants,
            regions: HashMap::new(),
            changed: HashSet::new(),
        }
    }
}
